tests/integration/docker.go

```diff
@@ -0,0 +1,144 @@
+	ContainerPort string
+	Proto         string
+}
+
+func (d dockerClient) runContainer(ctx context.Context, config *ContainerConfig) (*container.ContainerCreateCreatedBody, error) {
+	imageName, err := reference.ParseNormalizedNamed(config.image)
+	if err != nil {
+		return nil, fmt.Errorf("unable to normalize image name: %w", err)
+	}
+	fullName := imageName.String()
+
+	out, err := d.ImagePull(ctx, fullName, types.ImagePullOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to pull image: %w", err)
+	}
+	defer out.Close()
+
+	io.Copy(os.Stdout, out)
+
+	container, err := d.createNewContainer(ctx, fullName, config.ports, config.env, config.cmd)
+	if err != nil {
+		return nil, fmt.Errorf("unable create container: %w", err)
+	}
+	err = d.ContainerStart(ctx, container.ID, types.ContainerStartOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to start the container: %w", err)
+	}
+	fmt.Printf("container %s is started\n", container.ID)
+	return container, nil
+}
+
+func (d dockerClient) createNewContainer(ctx context.Context, image string, ports []*PortMapping, env []string, cmd []string) (*container.ContainerCreateCreatedBody, error) {
+	portBinding := nat.PortMap{}
+	for _, portmap := range ports {
+		hostBinding := nat.PortBinding{
+			//TODO: Allow for host ips to be specified
+			HostIP:   "0.0.0.0",
+			HostPort: portmap.HostPort,
+		}
+		containerPort, err := nat.NewPort("tcp", portmap.ContainerPort)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get the port: %w", err)
+		}
+		portBinding[containerPort] = []nat.PortBinding{hostBinding}
+	}
+
+	containerConfig := &container.Config{
+		Image: image,
+		Env:   env,
+	}
+	if len(cmd) > 0 {
+		containerConfig.Cmd = cmd
+	}
+	cont, err := d.ContainerCreate(
+		ctx,
+		containerConfig,
+		&container.HostConfig{
+			PortBindings: portBinding,
+		}, &network.NetworkingConfig{}, nil, "")
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create container: %w", err)
+	}
+	return &cont, nil
+}
+
+func (d dockerClient) removeContainer(ctx context.Context, id string) error {
+	fmt.Printf("container %s is stopping\n", id)
+	err := d.ContainerStop(ctx, id, &defaultTimeout)
+	if err != nil {
+		return fmt.Errorf("failed stopping container: %w", err)
+	}
+	fmt.Printf("container %s is stopped\n", id)
+
+	err = d.Client.ContainerRemove(ctx, id, types.ContainerRemoveOptions{
+		RemoveVolumes: true,
+		// RemoveLinks=true causes "Error response from daemon: Conflict, cannot
+		// remove the default name of the container"
+		RemoveLinks: false,
+		Force:       false,
+	})
+	if err != nil {
+		return fmt.Errorf("failed removing container: %w", err)
+	}
+	fmt.Printf("container %s is removed\n", id)
+	return nil
+}
+
+func (d dockerClient) printLogs(ctx context.Context, id string) {
+	out, err := d.ContainerLogs(ctx, id, types.ContainerLogsOptions{
+		ShowStdout: true,
+		ShowStderr: true,
+		Follow:     true,
+		Timestamps: true,
+	})
+	if err != nil {
+		panic(err)
```

Comment:
What was the reason for the panic here? For tests I don't mind passing `*testing.T` the using testify's `require.NoError`